server/src/course/course.recalgo.ts

```diff
@@ -0,0 +1,106 @@
+/**
+ * Applies stemming rules to reduce a word to its base form
+ */
+const stemWord = (word) => {
+  if (word.endsWith("sses")) {
+    return word.replace(/sses$/, 'ss');
+  } if (word.endsWith("ies")) {
+    return word.replace(/ies$/, 'y');
+  } if (word.endsWith("es") && !/[aeiou]es$/.test(word)) {
+    return word.replace(/es$/, '');
+  } if (word.endsWith("s") && word.length > 1 && !/[sxz]$/.test(word)) {
+    return word.replace(/s$/, '');
+  }
+  return word;
+}
+
+/**
+ * Preprocesses the description to remove pluralities and unnecessary punctuation
+ * @param description A course description that needs to be preprocessed
+ * @returns The processed description for a course
+ */
+export const preprocess = (description: string) => {
+  let sentences = description.match(/[^.!?]*[.!?]\s+[A-Z]/g) || [description];
+  let processedText = sentences.map(sentence => {
+    let words = sentence.match(/\b\w+\b/g) || [];
+    let cleanedWords = words.map(word => {
+      const singularWord = stemWord(word.toLowerCase());
+      return singularWord.replace(/[^\w\s]/g, '');
+    });
+    return cleanedWords.join(' ');
+  });
+  return processedText.join('. ');
+}
+
+/**
+ * Calculates the inverse document frequency for the given terms
+ * @param terms list of terms in the course description
+ * @param words list of all course descriptions as word arrays
+ * @returns a dictionary with terms as keys and their IDF scores as values
+ */
+export const idf = (terms, words) => {
+  let df = {};
+  let idf = {};
+  for (const term of terms) {
+    df[term] = words.reduce((count, wordsSet) => (count + (wordsSet.includes(term) ? 1 : 0)), 0);
+    idf[term] = 1 / (df[term] + 1);
+  }
+  return idf;
+}
+
+/**
+ * Calculates the TF-IDF vector for the given terms
+ * @param terms list of terms in the course description
+ * @param idf inverse document frequency (IDF) for the terms
+ * @returns a dictionary with terms as keys and their TF-IDF scores as values
+ */
+export const tfidf = (terms, idf) => {
+  let d = {};
+  for (const term of terms) {
+    if (!d[term]) {
+      d[term] = 0;
+    }
+    d[term]++;
+  }
+  for (const term in d) {
+    if (idf && idf[term] === undefined) {
+      idf[term] = 1;
+    }
+    d[term] *= idf[term];
```

Comment:
Maybe you could also normalize by dividing by term frequency here to make sure that the tfidf score accounts for different lengths of the documents to reflect an accurate importance for each term no matter document length.